File: test/Closet.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <algorithm>
#include <fstream>

using namespace std;

// Define a struct to hold the clothing details
struct Clothing {
    string name;
    string ID;
    vector<string> details;

    // Function to print clothing details
    void print() const {
        cout << "Clothing ID: " << ID << endl;
        cout << "Clothing Name: " << name << endl;
        cout << "Tags: " << endl;
        for (const auto& detail : details) {
            cout << "- " << detail << endl;
        }
        cout << endl;
    }

    // Function to check if any detail contains the search term
    bool contains(const string& term) const {
        if (name.find(term) != string::npos) {
            return true;
        }
        for (const auto& detail : details) {
            if (detail.find(term) != string::npos) {
                return true;
            }
        }
        return false;
    }
};

// Structure to input clothing details from the user
Clothing inputClothing() {
    Clothing Clothing;
    string detail;

    cout << "Enter Clothing ID: ";
    getline(cin, Clothing.ID);

    cout << "Enter Clothing Name: ";
    getline(cin, Clothing.name);
    //Adds details to the clothing
    while (true) {
        cout << "Enter Detail (or type 'done' to finish): ";
        getline(cin, detail);
        if (detail == "done") {
            break;
        }
        Clothing.details.push_back(detail);
    }

    return Clothing;
}

// Remove clothing from closet
void removeClothes(vector<Clothing>& closet){
    string ChosenID;
    cout << "Enter the Clothing ID that you want to remove" << endl;
    getline(cin, ChosenID);
    
    auto it = find_if(closet.begin(), closet.end(), [&](const Clothing& Clothing){
        return Clothing.ID ==
         ChosenID;
    });

    if (it != closet.end()){
        closet.erase(it);
        cout << "Clothing removed successfully." << endl;
    }else{
        cout << "There are no Clothings with the ID of " << ChosenID << endl;
    }
}
// Function to print all clothings in the closet
void printCloset(const vector<Clothing>& closet) {
    cout << "Closet contents:" << endl;
    for (size_t i = 0; i < closet.size(); ++i) {
        cout << "Clothing " << i + 1 << ":" << endl;
        closet[i].print();
    }
}

// Function to update clothing
void updateClothes(vector<Clothing>& closet){
    string ChosenID;
    string choice; 
    string detail_add;
    string detail_remove;
    cout << "Enter the clothing ID that you want to update" << endl;
    getline(cin, ChosenID);

    auto it = find_if(closet.begin(), closet.end(), [&](const Clothing& Clothing){
        return Clothing.ID == ChosenID;
    });
    //Checks to see if the inputted ID matches with a stored ID
    if (it != closet.end()){
        cout << "Clothing ID found" << endl;
        while(true){
        it->print();
        cout << "Do you want to 'add' or 'remove' details(Type 'exit' to exit)?" << endl;
        getline(cin, choice);
        //Adds additional details
        if(choice == "add"){
            while (true) {
            cout << "Enter detail (or type 'done' to finish): ";
            getline(cin, detail_add);
            if (detail_add == "done") {
                break;
                }
            it->details.push_back(detail_add);
            }
        //Removes existing details
        }else if (choice == "remove"){
            while (true) {
            cout << "Enter existing detail to remove (or type 'done' to finish): ";
            getline(cin, detail_remove);
            if (detail_remove == "done") {
                break;
                }
            
            auto detailIt = find(it->details.begin(), it->details.end(), detail_remove);
            if (detailIt != it->details.end()){
                it->details.erase(detailIt);
                cout << "Detail removed successfully. " << endl;
            } else{
                cout << "Detail not found." << endl;
                }
            }
        }else if (choice == "exit"){
            break;
        }else{
            cout << "Invalid choice. Please type 'add' or 'remove' " << endl;
        }
    }
    }else{
        cout << "Clothing ID not found." << endl;
    }
 }
// Search Function
void SearchClothes(const vector<Clothing>& closet){
    // Prompt user to enter a search term
    string searchTerm;
    cout << "Enter a detail to search for: ";
    getline(cin, searchTerm);
    cout << "Current Filters: " << searchTerm << endl;
    // Search and print clothing that contain the search term
    bool found = false;
    for (const auto& Clothing : closet) {
        if (Clothing.contains(searchTerm)) {
            Clothing.print();
            found = true;
        }
    }

    if (!found) {
        cout << "No clothings found with the detail: " << searchTerm << endl;
    }
}

// Function to save the closet to a file
void saveCloset(const vector<Clothing>& closet, const string& filename) {
    ofstream outFile(filename);

    if (!outFile) {
        cout << "Error opening text file!" << endl;
        return;
    }

    for (const auto& clothes : closet) {
        outFile << "Clothing ID: " << clothes.ID << endl;
        outFile << "Clothing Name: " << clothes.name << endl;
        for (const auto& detail : clothes.details) {
            outFile << "- " << detail << endl;
        }
    }

    outFile.close();
    cout << "Closet saved to " << filename << endl;
}

// Function to load the closet from a file in the same format as it was saved
void loadCloset(vector<Clothing>& closet, const string& filename) {
    ifstream inFile(filename);

    if (!inFile) {
        cout << "Error opening file for loading!" << endl;
        return;
    }

    closet.clear();
    Clothing clothes;
    string line;

    while (getline(inFile, line)) {
        if (line.find("Clothing ID:") != string::npos) {
            if (!clothes.name.empty()) {
                closet.push_back(clothes);  // Save the previous clothing item before starting a new one
            }

            clothes = Clothing();  // Reset the clothing object for the new item
            clothes.ID = line.substr(line.find(":") + 2);
        }
        else if (line.find("Clothing Name:") != string::npos) {
            clothes.name = line.substr(line.find(":") + 2);
        }
        else if (line.find("- ") == 0) {
            clothes.details.push_back(line.substr(2));  // Add the detail to the list
        }
    }

    if (!clothes.name.empty()) {
        closet.push_back(clothes);  // Save the last clothing item
    }

    inFile.close();
    cout << "Closet loaded from " << filename << endl;
}

int main() {
    // Create a vector to store the Clothings
    vector<Clothing> closet;
    
    string command;
    //Menu for Closet Manager
    cout << "Welcome to the Closet Manager" << endl;

    while (true) {
        cout << "Please choose an option" << endl;
        cout << "1. Add clothing" << endl;
        cout << "2. Remove clothing" << endl;
        cout << "3. View current closet" << endl;
        cout << "4. Update current clothing" << endl;
        cout << "5. Search for specific clothing" << endl;
        cout << "6. Save closet information to a text file" << endl;
        cout << "7. Load closet information from a text file" << endl;
        cout << "8. Exit" << endl;
        cout << "Select options from 1 to 8: ";
        getline(cin, command);
        
        if (command == "done") {
            break;
        } else if (command == "1") {
            closet.push_back(inputClothing());
        } else if (command == "2") {
            removeClothes(closet);    
        } else if (command == "3") {
            printCloset(closet);
        } else if (command == "4") {
            updateClothes(closet);    
        } else if (command == "5"){
            SearchClothes(closet);
        } else if (command == "6"){
            string filename;
            cout << "Enter file name to save to: ";
            getline(cin, filename);
            saveCloset(closet, filename);
        }else if (command == "7"){
            string filename;
            cout << "Enter file name to load from: ";
            getline(cin, filename);
            loadCloset(closet, filename);
        }else if (command == "8"){
            break;
        } else {
            cout << "Invalid command. Select a valid number." << endl;
        }
    }
    return 0;
}

File: test/Closet.py
from input_clothing import input_clothing
from remove_clothes import remove_clothes
from print_closet import print_closet
from update_clothes import update_clothes
from search_clothes import search_clothes
from switch_ids import switch_ids
from save_closet import save_closet
from load_closet import load_closet
from sort_by_color import sort_by_color
from checking_system import checking_system
from outfits import create_outfit
from outfits import add_clothing_to_outfit
from outfits import print_outfits


def main():
    i = 1

    closet = []
    outfits = []
    
    print("Welcome to the Closet Manager")

    while True:
        print("Please choose an option")
        print("1. Add clothing")
        print("2. Remove clothing")
        print("3. View current closet")
        print("4. Update current clothing")
        print("5. Search for specific clothing")
        print("6. Swap clothing")
        print("7. Save closet information to a text file")
        print("8. Load closet information from a text file")
        print("9. Sort closet by color")
        print("10. Check in clothing")
        print("11. Check out clothing")
        print("12. Take a picture of a clothing")
        print("13. Add an outfit")
        print("14. Add clothing items to an outfit")
        print("15. Print all outfits")
        print("16. Exit")
        command = input("Select options from 1 to 16: ")
        
        if command == "done":
            break
        elif command == "1":
            closet.append(input_clothing(closet, i))
            i += 1
        elif command == "2":
            remove_clothes(closet)    
        elif command == "3":
            print_closet(closet)
        elif command == "4":
            update_clothes(closet)    
        elif command == "5":
            search_clothes(closet)
        elif command == "6":
            switch_ids(closet)
        elif command == "7":
            filename = input("Enter file name to save to: ")
            save_closet(closet, outfits, filename)
        elif command == "8":
            filename = input("Enter file name to load from: ")
            load_closet(closet, outfits, filename)
        elif command == "9":
            sort_by_color(closet)
        elif command == "10":
            checking_system(closet, True)
        elif command == "11":
            checking_system(closet, False)
        elif command == "12":
            print("Camera not available")
        elif command == "13":
            outfit_name = input("Enter the name of the new outfit: ")
            create_outfit(outfits, outfit_name)
        elif command == "14":
           if not outfits:
            print("No outfits registered yet. Please create an outfit first.")
           else:
            print("Registered outfits: ")
            for index, outfit in enumerate(outfits, 1):
                print(f"{index}. {outfit.outfit_name}")
            outfit_name = input("Choose which outfit to add clothes too: ")
            chosen_id = input("Enter the Clothing ID that you want to add: ")
            clothing_type = input("Enter the type of clothing (Shirt, Pants, etc.): ")
            clothing_item = next((c for c in closet if c.ID == chosen_id), None)
            if clothing_item:
                add_clothing_to_outfit(outfits, outfit_name, clothing_type, clothing_item)
            else:
                print(f"No clothing item found with ID {chosen_id}")
        elif command == "15":
            print_outfits(outfits)
        elif command == "16":
            break
        else:
            print("Invalid command. Select a valid number.")
    print("\nExiting program.")

if __name__ == "__main__":
    main()